File: MAY 24/19 May/875. Koko Eating Bananas LC.cpp
class Solution {
    private:
    int check(vector<int> &piles, int speed)
    {
        int n = piles.size();
        long long hours=0;
        for(int i=0; i<n; i++)
            hours+=((0ll+piles[i]+speed-1)/speed);
        return hours;
    }
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int len = piles.size();
        long long left = 1, right=*max_element(piles.begin(), piles.end());

        long long minSpeed, speed;
        while(left<right)
        {
            speed = (left+right)/2;
            if(check(piles, speed)<=h)
                right = speed;
            else
                left = speed+1;
        }
        return left;
    }
};

File: APRIL 2025/AMZN/875. Koko Eating Bananas LC.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        long left = 1L, right = 1L*Arrays.stream(piles).max().getAsInt();
        long time, speed, N = piles.length;
        while(left<right) {
            speed = (left+right)/2;
            time = 0;
            for(int i=0; i<N; i++) {
                time += (piles[i]+speed-1)/speed;
            }
            if(time>h)
                left = speed+1;
            else
                right = speed;
        }
        return (int)right;
    }
}
